### ISSUE TITLE
Actual recaptha implementation, without 3rd party services

### ISSUE THREAD

**Issue by samip5** [DangerOnTheRanger/maniwani#130]
I would like to ask why is there no actual captha implementation that does not rely on on the wierd anime avatar thing or recaptha?
I don't want to use recaptha, so I'll be implementing or trying to an actual captha.

**Comment by DangerOnTheRanger** [comment]
reCAPTCHA and captchouli (at least for now) have no widespread automated solvers out there, and writing/maintaining a solid, readable CAPTCHA from scratch just honestly wasn't something I was/am personally interested in. I'd accept a PR for this, but I'm curious why you wouldn't want to go with captchouli.

**Comment by samip5** [comment]
I have no clue how to use it when I'm not that familiar with specific anime
characters and when it asks to check specific avatars of that anime
character.

>

**Comment by DangerOnTheRanger** [comment]
Fair enough - for the record, captchouli lets you configure [what characters to use](https://github.com/bakape/captchouli/blob/master/cmd/captchouli/main.go#L22), in case that helps; docs on how to pass flags to Dockerized captchouli are available [here](https://github.com/DangerOnTheRanger/maniwani/tree/master/build-helpers/captchouli).

**Comment by samip5** [comment]
Goal is something like this:
![captcha3](https://user-images.githubusercontent.com/1703002/70444371-71883f00-1aa2-11ea-85f0-93cee0c7708c.png)

Which captchouli cannot give me.

**Comment by samip5** [comment]
Basically the problem lies with the lack of knowledge on anime, period.

**Comment by DangerOnTheRanger** [comment]
@samip5 I'm still interested in merging this if you're still interested in working on it - if you have any questions about Maniwani's codebase/architecture I'll do my best to help out.

**Comment by samip5** [comment]
@DangerOnTheRanger I'm still interested in doing this, but as i have drafted in the pull request, I still need to find a library to help me out with it. The problem lies with the fact that Flask has very little library's for Captcha.

Do you know if we can use C++ native extension for it?
https://gitgud.io/LynxChan/NativeIM

**Comment by DangerOnTheRanger** [comment]
@samip5 Using C++ would be fine with me, though I am slightly concerned about how many more steps it would add to the build process (installing a C/C++ compiler, installing the necessary libraries, etc.). Docker would hide most of that complexity, though.

Reading the source code for LynxChan, it looks like its captcha code calls out to ImageMagick to write text on an image and then adds some boxes/circles, some distortion and blur, and so on. That should be possible in about as many lines of pure Python via [Pillow](https://pillow.readthedocs.io/en/stable/), I think, so that's another option.

**Comment by samip5** [comment]
It appears that the maintainer of LynxChan is not a fan of Python so they will most likely not help in any way at least in the implementation. Do you happen to be in the Python Discord server? (https://pythondiscord.com/)

https://balkanchan.ga/lynx/res/480.html

**Comment by DangerOnTheRanger** [comment]
I'm in it now, same username as here. I'm also on #maniwani on rizon.net at most US times of the day if IRC is more convenient.

**Comment by samip5** [comment]
I have been quite busy at the moment so that's why there's been no progress on this.

**Comment by DangerOnTheRanger** [comment]
No worries, we're not operating under deadlines - it's a hobby project after all ;)
